Bundle msal-common into msal-node

konstantin-msft opened the pull request.

konstantin-msft committed:
- Bundle msal-common into msal-node.

konstantin-msft committed:
Change files

konstantin-msft commented on lib/msal-node/tsconfig.json:
```diff
@@ -19,21 +19,25 @@
     "alwaysStrict": true,
     "noUnusedLocals": true,
-    "noUnusedParameters": true,
+    "noUnusedParameters": false,
```
Had to switch to false due to the high number of unused params in `msal-common`.

sameerag replied:
Can we please file an ADO to tackle this later?

konstantin-msft replied:
done

konstantin-msft commented on lib/msal-node/tsconfig.json:
```diff
@@ -19,21 +19,25 @@
+    "experimentalDecorators": true,
+    "baseUrl": ".",
+    "paths": {
+        "@azure/msal-common": ["src/msal-common"]
```
Create an alias for bundled `msal-common` to avoid changing references across the module.

konstantin-msft commented on lib/msal-node/package.json:
```diff
@@ -57,14 +56,14 @@
     "jest": "^29.5.0",
     "prettier": "2.8.7",
     "rollup": "^3.20.1",
+    "rollup-plugin-typescript-paths": "^1.4.0",
```
Rollup plugin to properly resolve `tsconfig.compileroptions.paths`.

konstantin-msft committed:
- Add `shx` package.

tnorling commented on lib/msal-node/tsconfig.json:
```diff
@@ -19,21 +19,25 @@
   "references": [
     {
-      "path": "../msal-common"
+      "path": "src/msal-common"
```
We probably should leave this unchanged. This is what makes "go to definition" work between libs.

konstantin-msft committed:
- Update e2e workflow.

konstantin-msft committed:
- Add bundledDependencies.

konstantin-msft committed:
- Add eslint-import-resolver-typescript package.

konstantin-msft committed:
- Add shx package.

konstantin-msft committed:
- Update e2e config.

konstantin-msft committed:
- Update e2e config.

sameerag reviewed (approved):
Lgtm. Would like to check if `noUsedParameters:true` can be fixed eventually.

konstantin-msft committed:
- Update e2e config.

konstantin-msft committed:
- Update e2e config.

konstantin-msft committed:
- Update e2e config.

konstantin-msft committed:
- Replace `tsconfig.compilerOptions.paths` with relative paths.

konstantin-msft committed:
- Remove `paths` from `tsconfig`.

konstantin-msft committed:
- Add "eslint-import-resolver-typescript" package.

konstantin-msft committed:
- Revert `jest.config.js`.

konstantin-msft commented:
@peterzenz @sameerag @tnorling I removed the original path alias for `msal-common` and replaced it with relative paths across the `msal-node` module. I didn't manage to make [jest moduleNameMapper](https://jestjs.io/docs/configuration#modulenamemapper-objectstring-string--arraystring) work for the e2e tests. Jest module name mapper doesn't work well for tests/utilities located outside of the `msal-node` that refer to the module directly, such as [this one](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/samples/e2eTestUtils/NodeCacheTestUtils.ts#L7).  Please let me know if you have any objections.

tnorling commented:
@konstantin-msft Did we consider [`bundleDepedencies`](https://docs.npmjs.com/cli/v8/configuring-npm/package-json#bundledependencies) or a similar approach of simply bundling msal-common into its dependents? I have some concerns about the ease of use/maintainability of this approach of copying the source...

In any case I think it would be good to write up our options and discuss pros/cons before we land on something

konstantin-msft commented:
> @konstantin-msft Did we consider [`bundleDepedencies`](https://docs.npmjs.com/cli/v8/configuring-npm/package-json#bundledependencies) or a similar approach of simply bundling msal-common into its dependents? I have some concerns about the ease of use/maintainability of this approach of copying the source...
> 
> In any case I think it would be good to write up our options and discuss pros/cons before we land on something

Bundling dependencies and specifying `@azure/msal-common` as a path alias was my initial approach.  However, it breaks e2e tests that import `msal-node` classes directly. For [example](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/samples/e2eTestUtils/NodeCacheTestUtils.ts#L7).

The options I can think of:
1. Bundling `msal-common` dependency and creating a path alias for it. Pros: less changes to `msal-node`, better maintainability. Cons: e2e tests won't work, requires changes to all jest config files.
2. Using relative references to `msal-common`. Pros: e2e tests work, no changes to jest config files required. Cons: worse maintainability because of `msal-common` being imported with relative paths.

tnorling commented:
> However, it breaks e2e tests that import `msal-node` classes directly. For [example](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/samples/e2eTestUtils/NodeCacheTestUtils.ts#L7).

Interesting...this should not have happened. We should be importing things from the package, not individual files/classes in samples and e2e tests. I'm also not sure I understand why our msal-common approach breaks node imports...anyway if this is the only blocker I would suggest we fix the e2e tests (we should fix this regardless of the approach we choose here IMO)

sameerag commented:
@konstantin-msft Can we start adding PR description to add better context?
